### PR TITLE
tipping

### PR DESCRIPTION
**Pull request by giuseppecrj**

### CODE REVIEW

File: contracts/src/spaces/facets/tipping/ITipping.sol [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+  event Tip(
+    uint256 indexed tokenId,
+    address indexed currency,
+    address sender,
+    address receiver,
+    uint256 amount
+  );
+
+  event TipMessage(bytes32 indexed messageId, bytes32 indexed channelId);
```
Comment (shuhuiluo) [inline contracts/src/spaces/facets/tipping/ITipping.sol:35]:
```suggestion
  event Tip(
    bytes32 indexed channelId,
    uint256 indexed tokenId,
    address indexed currency,
    address sender,
    address receiver,
    uint256 amount,
    bytes32 messageId
  );
```
Is there a benefit of indexing `messageId`?

File: contracts/src/spaces/facets/tipping/ITipping.sol [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  /// @dev Requires sender and receiver to be members of the space
+  /// @dev Requires amount > 0 and valid currency address
+  /// @dev Emits Tip and TipMessage events
+  function tip(TipRequest calldata tipRequest) external payable;
+
+  /// @notice Gets the total tips received for a token ID in a specific currency
+  /// @param tokenId The token ID to get tips for
+  /// @param currency The currency address to get tips in
+  /// @return The total amount of tips received in the specified currency
+  function tipsByCurrencyByTokenId(
```
Comment (shuhuiluo) [inline contracts/src/spaces/facets/tipping/ITipping.sol:61]:
```suggestion
  function tipsByCurrencyAndTokenId(
```
nit

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

File: contracts/src/spaces/facets/tipping/TippingFacet.sol [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import {Facet} from "contracts/src/diamond/facets/Facet.sol";
+
+contract TippingFacet is ITipping, ERC721ABase, Facet {
+  function __Tipping_init() external onlyInitializing {
+    _addInterface(type(ITipping).interfaceId);
+  }
+
+  /// @inheritdoc ITipping
+  function tip(TipRequest calldata tipRequest) external payable {
+    address sender = msg.sender;
```
Comment (shuhuiluo) [inline contracts/src/spaces/facets/tipping/TippingFacet.sol:21]: [`msg.sender`](https://www.evm.codes/?fork=cancun#a0) costs 2 gas which is cheaper than pushing it to the stack and moving it on the stack.

File: contracts/src/spaces/facets/tipping/TippingFacet.sol [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  function _validateTipRequest(
+    address sender,
+    address receiver,
+    address currency,
+    uint256 amount
+  ) internal view {
+    if (currency == address(0)) revert CurrencyIsZero();
+    if (sender == receiver) revert CannotTipSelf();
+    if (amount == 0) revert AmountIsZero();
+    if (_balanceOf(sender) == 0) revert SenderIsNotMember();
```
Comment (shuhuiluo) [inline contracts/src/spaces/facets/tipping/TippingFacet.sol:76]: Consider `CustomRevert`.

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

File: contracts/src/spaces/facets/tipping/ITipping.sol (hunk 1 above)
Comment (giuseppecrj) [reply contracts/src/spaces/facets/tipping/ITipping.sol:35]: this is for River stream.

**Review by shuhuiluo — APPROVED** [review]

**Pull request merged**